[AiLab] new db table to save machine learning model details

Erin007 opened the pull request:
Follow up to #37811 

Creates a new `UserMlModel` table and adds a row to it for each saved machine learning model. 

<img width="1185" alt="Screen Shot 2020-11-17 at 12 40 08 PM" src="https://user-images.githubusercontent.com/12300669/99427279-3659d700-28d3-11eb-95a9-f35b75a76c4c.png">

This is the initial work to allow us to use a trained machine learning model in the context of App Lab.

breville commented on dashboard/db/schema.rb:
```diff
@@ -28,7 +28,7 @@
   end
 
-  create_table "activity_sections", id: :integer, force: :cascade, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci" do |t|
+  create_table "activity_sections", id: :integer, force: :cascade, options: "ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci" do |t|
```
I'm not sure these changes are desired?

breville commented on dashboard/app/models/user_ml_model.rb:
```diff
@@ -0,0 +1,14 @@
+#
+# Table name: user_ml_models
+#
+#  id         :bigint           not null, primary key
```
Most of our other models seem to use `:integer`?

sureshc replied:
We've bumped up against the maximum SQL integer limit on a few tables, so starting out with bigint seems appropriate.

sureshc commented:
Could we isolate the rails database migration (and the associated empty model and the new schema.rb) to its own Pull Request? That's been our convention in the past.

sureshc commented on dashboard/app/models/user_ml_model.rb:
```diff
@@ -0,0 +1,14 @@
+#
+#  id         :bigint           not null, primary key
+#  user_id    :integer
+#  model_id   :string(255)
```
Do we need an index on model_id or any other column other than the primary key?

Erin007 commented:
> Could we isolate the rails database migration (and the associated empty model and the new schema.rb) to its own Pull Request? That's been our convention in the past.

Yes! Thanks for the reminder.

Erin007 commented:
ok, @sureshc and @breville I've tidied this PR based on your suggestions, can you take another look please? Thank you!

breville commented on dashboard/test/factories/user_ml_models.rb:
```diff
@@ -0,0 +1,18 @@
+# == Schema Information
```
We have very few files in this directory.  It might be worth holding off adding this file until we know we will use it.

breville commented on dashboard/test/models/user_ml_model_test.rb:
```diff
@@ -0,0 +1,7 @@
+require 'test_helper'
```
It might also be good to hold off on adding this file until we know we are going to use it.

sureshc commented:
Will we frequently be looking up all user models owned by a specific user? Or all user models for a specific model id? If so, we may want an index on each of those columns. Are we going to support soft deletes of models? Should we have a deleted_at column? Could models contain personal data? If so, will we need a purged_at column as well to identify when sensitive information in the user model has been de-identified?

Erin007 commented:
> Will we frequently be looking up all user models owned by a specific user? Or all user models for a specific model id? If so, we may want an index on each of those columns. Are we going to support soft deletes of models? Should we have a deleted_at column? Could models contain personal data? If so, will we need a purged_at column as well to identify when sensitive information in the user model has been de-identified?

Yes, we will be looking up ml models by model_id and user_id so I added an index for each of those. Also, added deleted_at and purged_at columns just in case.  Hoping we can prevent users from saving models that have sensitive information, but good to be ready to purge/delete if needed. Thanks!

Erin007 commented:
@sureshc any other requested revisions on this one? Thanks!

sureshc commented on dashboard/db/migrate/20201117160845_create_user_ml_models.rb:
```diff
@@ -0,0 +1,15 @@
+class CreateUserMlModels < ActiveRecord::Migration[5.1]
+  def change
+    create_table :user_ml_models do |t|
```
Currently the production database server is configured with `utf8` as the default character set and collation, and because this migration does not explicitly specify the character set, this table will get created with `utf8`, but your local development database server has the default setting `utf8mb4` configured on the `dashboard` database so there's a diff between the `schema.rb` being committed and what will be generated on `production`. The `staging` database was rebuilt recently, also has 'utf8mb4' now as it's default, and is the authoritative creator/committer of `schema.rb`. @wjordan is working on updating the production configuration to eliminate this, but in the short term, we may want to explicitly specify the character set in this migration. Looking up how to do that ...

sureshc replied:
[Here's an example](https://github.com/code-dot-org/code-dot-org/blob/5975463436e6a99fa8b9f051606544a3056f4e20/dashboard/db/migrate/20200518181001_convert_foorm_submissions_to_utf8mb4.rb#L3) of how to change the default character set after the table has been created.  I don't see a way to specify this directly with the rails `create_tasble` method.

sureshc reviewed (commented):
> @sureshc any other requested revisions on this one? Thanks!

Sorry for not noticing this earlier, but I think we should figure out a way to resolve the character set discrepancy between `production` and `staging` before merging.
